content-resources/src/main/java/flyway/oskari/V1_46_9__replace_externalids_in_mapful_bglayerselection_plugin_config.java

```diff
@@ -0,0 +1,228 @@
+        Map<String, Integer> externalIdToLayerId = new HashMap<>();
+
+        String sql = "SELECT maplayerid, externalid FROM oskari_maplayer_externalid";
+        try (PreparedStatement ps = conn.prepareStatement(sql);
+                ResultSet rs = ps.executeQuery()) {
+            while (rs.next()) {
+                int layerId = rs.getInt("maplayerid");
+                String externalId = rs.getString("externalid");
+                LOG.debug("externalId:", externalId, "layerId:", layerId);
+                externalIdToLayerId.put(externalId, layerId);
+            }
+        }
+
+        return externalIdToLayerId;
+    }
+
+    private List<BundleConfigNState> getBundleConfigs(Connection conn, int mapfullBundleId) throws SQLException {
+        List<BundleConfigNState> configs = new ArrayList<>();
+
+        String sql = "SELECT view_id, seqno, config, state FROM portti_view_bundle_seq WHERE bundle_id = ?";
+        try (PreparedStatement statement = conn.prepareStatement(sql)) {
+            statement.setInt(1, mapfullBundleId);
+            try (ResultSet rs = statement.executeQuery()) {
+                while (rs.next()) {
+                    BundleConfigNState config = new BundleConfigNState();
+                    config.viewId = rs.getInt("view_id");
+                    config.seqNo = rs.getInt("seqno");
+                    config.config = JSONHelper.createJSONObject(rs.getString("config"));
+                    config.state = JSONHelper.createJSONObject(rs.getString("state"));
+                    configs.add(config);
+                }
+            }
+        }
+
+        return configs;
+    }
+
+    protected static List<BundleConfigNState> getBundleConfigsToUpdate(List<BundleConfigNState> bundleConfigs,
+            Map<String, Integer> externalIdToLayerId) {
+        List<BundleConfigNState> toUpdate = new ArrayList<>();
+
+        for (BundleConfigNState bundleConfig : bundleConfigs) {
+            boolean updatedConfig = updateConfig(bundleConfig.config, externalIdToLayerId);
+            boolean updatedState = updateState(bundleConfig.state, externalIdToLayerId);
+            if (updatedConfig || updatedState) {
+                toUpdate.add(bundleConfig);
+            }
+        }
+
+        return toUpdate;
+    }
+
+    protected static boolean updateConfig(JSONObject config, Map<String, Integer> externalIdToLayerId) {
+        if (config == null) {
+            return false;
+        }
+        try {
+            JSONArray plugins = config.optJSONArray("plugins");
+            JSONObject bgPlugin = findBGPlugin(plugins);
+            if (bgPlugin == null) {
+                return false;
+            }
+            JSONObject bgPluginConfig = bgPlugin.optJSONObject("config");
+            if (bgPluginConfig == null) {
+                return false;
+            }
+            JSONArray baseLayers = bgPluginConfig.optJSONArray("baseLayers");
+            return replaceExternalIds(baseLayers, externalIdToLayerId);
```

Comment:
In addition to Oskari.mapframework.bundle.mapmodule.plugin.BackgroundLayerSelectionPlugin also Oskari.mapframework.bundle.mapmodule.plugin.LayerSelectionPlugin might have layer references in the config:

- conf.defaultBaseLayer with a value of a single layer id
- conf.baseLayers with a value of array of layer ids

These need to be migrated as well.

Reply:
Let's do another PR for this to get this show on the road.